### PR TITLE
Stripe Integration

### LINKED ISSUE #12: Billing & subscriptions

### PR DESCRIPTION
**Pull request by ukrocks007**
This PR adds Stripe Integration to the saas-starter-kit. Fixes #12 

How to test:
 - Create an account on Stripe.
 - Get the `Secret Key` from [API Keys](https://dashboard.stripe.com/test/apikeys) section.
 - Install Stripe Cli and authenticate
 - Create products and prices using the Stripe dashboard
 - Run `npm run sync-stripe` to ensure db & stripe have the same set of active products and prices.
 - Run `stripe listen --forward-to=localhost:4002/api/webhooks/stripe` in a terminal to forward the webhook events to local apis
 - Run the app with `npm run dev`
 - Test Credit Card number: `4242 4242 4242 4242`

### CODE REVIEW

File: lib/env.ts [hunk 1]
```diff
@@ -107,6 +108,11 @@ const env = {
   maxLoginAttempts: Number(process.env.MAX_LOGIN_ATTEMPTS) || 5,
 
   slackWebhookUrl: process.env.SLACK_WEBHOOK_URL,
+
+  stripe: {
+    stripeSecretKey: process.env.STRIPE_SECRET_KEY,
+    stripeWebhookSecret: process.env.STRIPE_WEBHOOK_SECRET,
```
Comment (devkiran) [inline lib/env.ts:114]:
```suggestion
    secretKey: process.env.STRIPE_SECRET_KEY,
    webhookSecret: process.env.STRIPE_WEBHOOK_SECRET,
```

File: lib/stripe.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  apiVersion: '2022-11-15',
+  // Register this as an official Stripe plugin.
+  // https://stripe.com/docs/building-plugins#setappinfo
+  appInfo: {
+    name: 'saas-starter-kit',
+    version: '0.1.0',
+  },
+});
+
+export const getURL = () => {
```
Comment (devkiran) [inline lib/stripe.ts:16]: Can we directly use the existing `appUrl`, the value would be `https` on production.

File: models/stripeTeam.ts
Comment (devkiran) [inline models/stripeTeam.ts:1]: The table `StripeTeam` might be unnecessary. Can we add `stripeCustomerId` to `Team` instead? I assume we never want to have multiple Stripe customer id per team.

File: package.json [hunk 3]
```diff
@@ -27,13 +28,15 @@
...
     "@retracedhq/logs-viewer": "2.7.0",
+    "@stripe/stripe-js": "1.54.2",
     "@retracedhq/retraced": "0.7.6",
     "@sentry/nextjs": "7.92.0",
     "@tailwindcss/typography": "0.5.10",
     "bcryptjs": "2.4.3",
     "classnames": "2.5.1",
     "cookies-next": "4.1.0",
     "crypto-js": "4.2.0",
+    "dotenv": "16.3.1",
```
Comment (devkiran) [inline package.json:39]: Node `v20.6.0` has built-in support for `.env` file. We can use `dotenv` if we want to support older versions of Node.

File: pages/api/teams/[slug]/payments/create-portal-link.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  const teamMember = await throwIfNoTeamAccess(req, res);
+  const session = await getSession(req, res);
+  const customerId = await getStripeCustomerId(teamMember, session);
+
+  const { url } = await stripe.billingPortal.sessions.create({
+    customer: customerId,
+    return_url: `${getURL()}teams/${teamMember.team.slug}/payments`,
+  });
+
+  res.json({ url });
```
Comment (devkiran) [inline pages/api/teams/[slug]/payments/create-portal-link.ts:40]: Can we wrap this within data object? `{ data: { url } }` to have consistent response format.

File: pages/teams/[slug]/payments.tsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        </>
+      )}
+    </>
+  );
+};
+
+export async function getServerSideProps({
+  locale,
+}: GetServerSidePropsContext) {
+  if (!env.teamFeatures.apiKey) {
```
Comment (devkiran) [inline pages/teams/[slug]/payments.tsx:136]:
```suggestion
  if (!env.teamFeatures.payments) {
```

**Review by devkiran — CHANGES_REQUESTED** [review]
Sharing some nitpicks. I haven't tested the integration yet. Need to configure Stripe first.

File: locales/en/common.json [hunk 6]
```diff
@@ -198,5 +198,11 @@
   "remove-other-browser-session-warning": "Once removed you will be logged out of the app on that browser.",
   "remove-current-browser-session-warning": "You are about to remove your current session. Once you do this, you will be logged out of the app.",
   "session-removed": "Browser session removed.",
-  "remove-team-restricted": "Please contact our support if you want to remove this team. We will help you with that."
+  "remove-team-restricted": "Please contact our support if you want to remove this team. We will help you with that.",
+  "open-customer-portal": "Open Customer Portal",
+  "subscriptions": "Subscriptions",
+  "from": "From",
+  "to": "To",
+  "stripe-checkout-fallback-error": "Somthing went wrong while initiating checkout, please try again later.",
```
Comment (devkiran) [inline locales/en/common.json:206]:
```suggestion
  "stripe-checkout-fallback-error": "Something went wrong while initiating checkout, please try again later.
",
```

File: components/team/TeamTab.tsx [hunk 7]
```diff
@@ -76,6 +77,18 @@ const TeamTab = ({ activeTab, team, heading, teamFeatures }: TeamTabProps) => {
     });
   }
 
+  if (
+    teamFeatures.payments &&
+    canAccess('team_payments', ['create', 'update', 'read', 'delete'])
+  ) {
+    navigations.push({
+      name: 'Payments',
```
Comment (devkiran) [inline components/team/TeamTab.tsx:85]: Calling this `Billing` might be more accurate.

File: package.json (hunk 3 above)
Comment (ukrocks007) [reply package.json:39]: had to convert ts file to js for this to work

File: prisma/schema.prisma
Comment (devkiran) [inline prisma/schema.prisma:1]: Migration files for schema changes are missing.

File: pages/teams/[slug]/billing.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+          <div className="">
+            <Button
+              type="button"
+              color="primary"
+              size="md"
+              loading={portalLinkLoading}
+              onClick={redirectToCustomerPortal}
+            >
+              {t('open-customer-portal')}
+            </Button>
```
Comment (devkiran) [inline pages/teams/[slug]/billing.tsx:89]: Should we hide the link to customer portal if there is no subscription yet or customer id created?

File: package.json [hunk 9]
```diff
@@ -27,6 +28,7 @@
     "@prisma/client": "5.7.1",
     "@react-email/components": "0.0.12",
     "@retracedhq/logs-viewer": "2.7.0",
+    "@stripe/stripe-js": "1.54.2",
```
Comment (devkiran) [inline package.json:31]: Looks like we're not using this package.

File: package.json [hunk 10]
```diff
@@ -82,6 +85,7 @@
     "prettier": "3.1.1",
     "prettier-plugin-tailwindcss": "0.5.11",
     "prisma": "5.7.1",
+    "ts-node": "10.9.2",
```
Comment (devkiran) [inline package.json:88]: Looks like we're not using this package.

File: pages/teams/[slug]/billing.tsx (hunk 8 above)
Comment (ukrocks007) [reply pages/teams/[slug]/billing.tsx:89]: When user clicks on the button to open customer portal, in the api call we check if customer is registered on stripe or not, if not we create a customer on stripe save the customerId and open portal

File: lib/env.ts [hunk 11]
```diff
@@ -96,6 +96,14 @@ const env = {
     webhook: process.env.FEATURE_TEAM_WEBHOOK === 'false' ? false : true,
     apiKey: process.env.FEATURE_TEAM_API_KEY === 'false' ? false : true,
     auditLog: process.env.FEATURE_TEAM_AUDIT_LOG === 'false' ? false : true,
+    payments:
+      process.env.FEATURE_TEAM_PAYMENTS === 'false'
+        ? false
+        : process.env.STRIPE_SECRET_KEY &&
+            process.env.STRIPE_WEBHOOK_SECRET &&
+            process.env.CRON_API_KEY
```
Comment (devkiran) [inline lib/env.ts:104]: `CRON_API_KEY` What is this for? Can't see it being used anywhere

File: lib/stripe.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+export const stripe = new Stripe(env.stripe.secretKey ?? '', {
+  // https://github.com/stripe/stripe-node#configuration
+  apiVersion: '2022-11-15',
+  // Register this as an official Stripe plugin.
+  // https://stripe.com/docs/building-plugins#setappinfo
+  appInfo: {
+    name: 'saas-starter-kit',
+    version: '0.1.0',
+  },
```
Comment (devkiran) [inline lib/stripe.ts:13]: is this really needed? This option is deprecated https://stripe.com/docs/building-plugins#setappinfo

File: lib/env.ts (hunk 11 above)
Comment (ukrocks007) [reply lib/env.ts:104]: using in pages/api/webhooks/payment.ts to validate webhook calls

File: lib/stripe.ts (hunk 12 above)
Comment (ukrocks007) [reply lib/stripe.ts:13]: fixed

File: lib/env.ts (hunk 11 above)
Comment (ukrocks007) [reply lib/env.ts]: removed

File: prisma/schema.prisma [hunk 13]
```diff
@@ -135,3 +136,39 @@ model ApiKey {
...
+  updatedAt   DateTime @default(now())
+  Price       Price[]
+}
+
+model Price {
+  id            String   @id @default(uuid())
+  billingScheme String
+  currency      String
+  planId        String
+  recurring     Json
```
Comment (devkiran) [inline prisma/schema.prisma:169]: The `recurring` field is a bit confusing, why can't we a `price` or `amount` field instead? and store everything else in `metadata`? `metadata` can be a JSON field, and store provider specific data. There should be a `amount` (could be 0) exists for every plans/price, so it should be a required field. What do you think?

Comment (ukrocks007) [reply prisma/schema.prisma:169]: There are different possible plans
 - Flat price
 - Tiered price
 - Usage based
 - Package based, etc.
So not every price will have the amount
Will create a nullable amount column & save recurring in metadata

**Comment by KazeroG** [comment]
There is some local test video : 

https://github.com/boxyhq/saas-starter-kit/assets/17680194/46481ba9-398c-447f-8f9f-76abf641a034


Feature to add :
- [ ] Good Documentation
- [ ] How to add which page is restricted based on subscription
- [ ] While registering the user, check if the user already exists in Stripe, if yes, sync it, if not, create the client.

**Review by KazeroG — COMMENTED** [review]
Seems good for me

**Review by KazeroG — COMMENTED** [review]
Seems Good for me

File: syncStripe.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    const [products, prices] = await Promise.all([
+      stripe.products.list({ active: true }),
+      stripe.prices.list({ active: true }),
+    ]);
+
+    if (prices.data.length > 0 && products.data.length > 0) {
+      await cleanup(prisma);
+
+      await seedServices(products.data, prisma);
+      await seedPrices(prices.data, prisma);
```
Comment (ukrocks007) [inline syncStripe.js:27]: Convert to a transaction

File: prisma/schema.prisma [hunk 15]
```diff
@@ -135,3 +136,42 @@ model ApiKey {
 
   team Team @relation(fields: [teamId], references: [id], onDelete: Cascade)
 }
+
+model Subscription {
+  id         String    @id @unique
```
Comment (ukrocks007) [inline prisma/schema.prisma]: remove `@unique`

**Review by deepakprabhakara — APPROVED** [review]

File: prisma/schema.prisma (hunk 15 above)
Comment (deepakprabhakara) [reply prisma/schema.prisma]: Waiting for this last task

Comment (ukrocks007) [reply prisma/schema.prisma]: Done with this @deepakprabhakara

**Review by deepakprabhakara — APPROVED** [review]

**Review by devkiran — APPROVED** [review]

**Review by deepakprabhakara — APPROVED** [review]

**Review by niwsa — APPROVED** [review]

**Pull request merged**